Add support for S3-compatible endpoints for confmap/s3provider (#47602)
open-telemetry/opentelemetry-collector-contrib: confmap/provider/s3provider/README.md

## Summary

> This package provides a `ConfigMapProvider` implementation for S3 (`s3provider`) that allows the Collector to load configuration by fetching config objects from Amazon S3 or any S3-compatible storage service.

Reviewer: ```suggestion
This package provides a `ConfigMapProvider` implementation for Amazon S3 (`s3provider`) that allows the Collector to load configuration by fetching config objects from Amazon S3 or any Amazon S3-compatible storage service.
```

I don't think we need to spell out "Amazon Simple Storage Service (Amazon S3)" on first use, but I do have to at least ask that we use the product name. :)

## How it works

- It will be called by `ConfigMapResolver` to load configuration for the Collector.
- By giving a config URI starting with prefix `s3://`, this `s3provider` will be used to download config objects from the given S3 URIs, and then use the downloaded configuration during Collector initialization.

## URI formats

> ### AWS S3 (virtual-hosted-style)

Reviewer: ```suggestion
### Amazon S3 (virtual-hosted-style)
```

```
s3://[BUCKET].s3.[REGION].amazonaws.com/[KEY]
```

Example:

```
s3://doc-example-bucket.s3.us-west-2.amazonaws.com/config.yaml
```

### S3-compatible services (path-style)

> For non-AWS S3-compatible services such as MinIO or DigitalOcean Spaces:

Reviewer: ```suggestion
### Amazon S3-compatible services (path-style)

For Amazon S3-compatible services such as MinIO or DigitalOcean Spaces:
```

```
s3://[ENDPOINT_HOST]/[BUCKET]/[KEY]?region=[REGION]
```

> - `ENDPOINT_HOST` — the hostname (and optional port) of the S3-compatible service

Reviewer: ```suggestion
- `ENDPOINT_HOST` — the hostname (and optional port) of the Amazon S3-compatible service
```
